src/sql/sqlops.proposed.d.ts

```diff
@@ -1365,4 +1365,620 @@ declare module 'sqlops' {
+		}
+
+		/**
+		 * Defines the contracts needed to manage the lifetime of a notebook server.
+		 */
+		export interface ServerManager {
+			/**
+			 * Indicates if the server is started at the current time
+			 */
+			readonly isStarted: boolean;
+
+			/**
+			 * Event sent when the server has started. This can be used to query
+			 * the manager for server settings
+			 */
+			readonly onServerStarted: vscode.Event<void>;
+
+			/**
+			 * Starts the server. Some server types may not support or require this.
+			 * Should no-op if server is already started
+			 */
+			startServer(): Thenable<void>;
+
+			/**
+			 * Stops the server. Some server types may not support or require this
+			 */
+			stopServer(): Thenable<void>;
+		}
+
+		//#region Content APIs
+		/**
+		 * Handles interacting with file and folder contents
+		 */
+		export interface ContentManager {
+			/* Reads contents from a Uri representing a local or remote notebook and returns a
+			 * JSON object containing the cells and metadata about the notebook
+			 */
+			getNotebookContents(path: string): Thenable<INotebook>;
+
+			/**
+			 * Save a file.
+			 *
+			 * @param path - The desired file path.
+			 *
+			 * @param notebook - notebook to be saved.
+			 *
+			 * @returns A thenable which resolves with the file content model when the
+			 *   file is saved.
+			 */
+			save(path: string, notebook: INotebook): Thenable<INotebook>;
+		}
+
+		export interface INotebook {
+
+			readonly cells: ICell[];
+			readonly metadata: INotebookMetadata;
+			readonly nbformat: number;
+			readonly nbformat_minor: number;
+		}
+
+		export interface INotebookMetadata {
+			kernelspec: IKernelInfo;
+			language_info?: ILanguageInfo;
+		}
+
+		export interface IKernelInfo {
+			name: string;
+			language?: string;
+			display_name?: string;
+		}
+
+		export interface ILanguageInfo {
+			name: string;
+			version: string;
+			mimetype?: string;
+			codemirror_mode?: string | ICodeMirrorMode;
+		}
+
+		export interface ICodeMirrorMode {
+			name: string;
+			version: string;
+		}
+
+		export interface ICell {
+			cell_type: CellType;
+			source: string | string[];
+			metadata: {
+				language?: string;
+			};
+			execution_count: number;
+			outputs?: ICellOutput[];
+		}
+
+		export type CellType = 'code' | 'markdown' | 'raw';
+
+		export interface ICellOutput {
+			output_type: OutputType;
+		}
+		export interface IStreamResult extends ICellOutput {
+			/**
+			 * Stream output field defining the stream name, for example stdout
+			 */
+			name: string;
+			/**
+			 * Stream output field defining the multiline stream text
+			 */
+			text: string | Buffer;
+		}
+		export interface IDisplayResult extends ICellOutput {
+			/**
+			 * Mime bundle expected to contain mime type -> contents mappings.
+			 * This is dynamic and is controlled by kernels, so cannot be more specific
+			 */
+			data: {};
+			/**
+			 * Optional metadata, also a mime bundle
+			 */
+			metadata?: {};
```

Comment:
Is there any way to get better types on these? Maybe a parameterized type like `IDisplayResult<T, U> { data: T, metadata: U }`? Not too important but for API users it's helpful to not have to worry about manually figuring out the types too much

Reply:
Good question. I can consider adding in some kind of base "this is a key value pair" type which I think JupyterLab does. I'll address these in follow up revisions since we're moving things as-is